### ISSUE TITLE
Simplify API

### ISSUE THREAD

**Issue by vadimdemedes** [novemberborn/cached-module-loader#8]
I think the API for that module could be simple and export only one factory function, which returns a modified `require()`:

``` js
const cachedRequire = require('cached-module-loader'){
  cachePath: join(__dirname, '.module-cache'),
  ignore: [/regex/, /to/, /ignore/]
});

const babel = cachedRequire('babel-core');
const got = cachedRequire('got');
```

`cachedRequire()`'s source would be something along:

``` js
function cachedRequire(moduleId) {
  if (!isCacheExists(moduleId)) {
    bundleDependencies(moduleId);
  }

  return loadInThisContext(moduleId);
}
```

What do you think?

**Comment by novemberborn** [comment]
Creating the bundle is asynchronous, and we'll have to expose data to help with cache validation and so forth. More of that logic could be integrated though.

I'm pretty happy with `load` actually, it signifies loading from the bundle rather than being an actual `require` implementation. I don't think it's too much of a problem that you need to pass an absolute path, working with these bundles should require you to know what you're doing. The trade-offs are too complex to hide behind a simple API.

**Comment by vadimdemedes** [comment]
> Creating the bundle is asynchronous, and we'll have to expose data to help with cache validation and so forth. More of that logic could be integrated though.

Good point, forgot about that. I guess `bundleDependencies()` must stay, although I'd shorten it to `bundle()`.

> I'm pretty happy with load actually, it signifies loading from the bundle rather than being an actual require implementation. 

Not sure if we're on the same page here. The API I proposed is for top-level usage, not inside bundles. `cachedRequire` is more intuitive (at least) to me, rather than `loadInThisContext`.

**Comment by vadimdemedes** [comment]
One more idea: I think if the cache is missing, we shouldn't build it right away. Taking AVA into account, we want tests to start running as soon as possible. If the cache is missing, native `require()` should be used, but let's build the cache after the tests are done. 

What do you think?

**Comment by vadimdemedes** [comment]
cc @sindresorhus @jamestalmage @sotojuan @jfmengels

**Comment by novemberborn** [comment]
> Good point, forgot about that. I guess bundleDependencies() must stay, although I'd shorten it to bundle().

👍 

> > I'm pretty happy with load actually, it signifies loading from the bundle rather than being an actual require implementation.
> 
> Not sure if we're on the same page here. The API I proposed is for top-level usage, not inside bundles. cachedRequire is more intuitive (at least) to me, rather than loadInThisContext.

Yes I understood that you meant the top-level API. I just think using the bundle requires more care than a "it's like require" API method implies. For one we should resolve the module identifier relative to the calling module.

That said, `loadInThisContext` is unnecessarily tied to the underlying `Script` usage, and providing the resolved module identifier as well as the user-friendly identifier is cumbersome. Given that we're already passing `module` we could simplify to:

``` js
load('babel-core', module)
```

and internally use `module.require.resolve('babel-core')` to get the filename. I think.

> One more idea: I think if the cache is missing, we shouldn't build it right away. Taking AVA into account, we want tests to start running as soon as possible. If the cache is missing, native require() should be used, but let's build the cache after the tests are done.

This is why I don't want the interface to be too `require`-like. You're going to have to make some trade-offs when you use this package.

Ideally we can find a good caching strategy (#4) we can ship with which will help here.

**Comment by sindresorhus** [comment]
> Given that we're already passing module we could simplify to: load('babel-core', module)

👍
